### PR TITLE
Add a pre-commit script to detect missing i18n implementations

### LINKED ISSUE #9423: Create python script to automatically detect potentially missing i18n strings
### Problem

A common source of feedback on code reviews is when folks miss i18n opportunities. With a few simple regular expressions, we can detect such places in our html files that need i18n, and then use pre-commit to run the script have it auto-fail a pull request if the i18n strings are missing!

Examples regular expressions:

In our html, all `>` must be followed either by:
- a newline
- a `<`
- or a `$_` or `$:_` or `$ungettext` or `$:ungettext`
- or `$`, denoting some sort of variable being rendered

all occurrences of the html attirbutes `(title|placeholder|alt)=['"]` must either be:
- empty; either followed by `['"]`
- Or followed by a `$_` or `$:_` or `$ungettext` or `$:ungettext`
- or `$`, denoting some sort of variable being rendered

These rules should be enough to catch most cases of missing i18n in our codebase. To allow a break-the-glass, anywhere we need the i18n things, we can also allow `$(`. E.g. `<p>$("String I don't want translated")</p>`. This script should output a warning in these cases, but not fail.

Example run:

```sh
$ PYTHONPATH=. python scripts/detect_missing_i18n.py openlibrary/templates/foo.html openlibrary/templates/bar.html
ERRO openlibrary/templates/foo.html:123:11         <p>Click here!</p>
WARN openlibrary/templates/foo.html:128:4  <p>$("PDF")</p>

2 files scanned. 1 error found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
```

We can then wire this up to pre-commit to auto run on changed html files.

### Is there a precedent of this approach succeeding elsewhere?
- See https://github.com/internetarchive/openlibrary/pull/8900/files#diff-63a9c44a44acf85fea213a857769990937107cf072831e1a26808cfde9d096b9 for an example of how to run a python script as a pre-commit

### Which suggestions or requirements should be considered for how feature needs to appear or be implemented?
- The script should provide output as its going; not process all the files and _then_ provide output
- The script should use `FnToCLI` to mange CLI parameters; eg..

```py
def main(files: list[Path]):
    """
    :param files: The html files to check for missing i18n
    """
    ...


if __name__ == "__main__"
    FnToCLI(main).run()
```

### Stakeholders
@rebecca-shoptaw 

**Note:** Before [making a new branch](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#making-changes-and-creating-a-pull-request) _or_ updating an existing one, please [ensure your branch is up to date](https://github.com/internetarchive/openlibrary/wiki/Git-Cheat-Sheet#working-on-your-branch).

### PR DESCRIPTION
**Pull request by pidgezero-one**
Closes #9423

- Adds a pre-commit script that finds HTML that might be missing i18n implementation

### Technical
Parses the contents of html files in specific directories to detect the line and column numbers for tags that might be missing i18n implementation.

Mostly a series of loops to ensure that the provided indexes are correct.

### Testing
I've been adding junk data to html files and running both `pre-commit run` and `pre-commit run --all-files`, and it looks like the scope of those two different commands is correct. However, the regex approach returns quite a lot of what I think are false positives.

### Stakeholders
@rebecca-shoptaw

### CODE REVIEW

**Comment by pidgezero-one** [comment]
@rebecca-shoptaw @cdrini 

So far this appears to be working as expected according to the issue description, however my regex approach flags quite a lot of what I think are false positives (example: https://results.pre-commit.ci/run/github/69609/1718241262.NJ_4UJGwTqOunVrs7ivFkA)

Some examples:
- I think this should ignore `<code>` elements.
- Should it be considered an error when the contents of an element are a template variable? Example: `ERRO openlibrary/macros/SearchResults.html:23:36  <a href="$book.key/$title.replace(' ', '-')">$book.title_prefix $title</a></span>` or `ERRO openlibrary/templates/covers/book_cover_single_edition.html:18:16  <div class="BookTitle">$:macros.TruncateString(title, 70)`
- I'm not sure what the meaning is of `\$$`, for example: `ERRO openlibrary/templates/admin/sponsorship.html:69:6  <td>\$$('{:,.2f}'.format(summary['avg_scan_cost'] / 100.))</td>` - is this something else that should avoid flagging?
- There are some instances where the same error is flagged twice when an outer element is on the same line as an inner element and they're separated by a superfluous space:
```
ERRO openlibrary/templates/subjects.html:79:12  <span class="count">        <a href="/search?${page.subject_type}_facet=$page.name.replace('&','%26')">Search within $page.name</a></span>
ERRO openlibrary/templates/subjects.html:79:40  <a href="/search?${page.subject_type}_facet=$page.name.replace('&','%26')">Search within $page.name</a></span>
```
(I am assuming that this is an indication that the space should be deleted and not that the script should account for this?)
- I don't have a good answer for tags that are valid inside i18n functions, like <strong> or <em> or maybe <span>. I thought about ignoring any result that matches those three tags, but then that would ignore things like `<div><span>untranslated text</span></div>`.

Please let me know what your thoughts are on the above and I'll be happy to make any changes!

**Comment by rebecca-shoptaw** [comment]
@pidgezero-one This is looking great!!
To respond in order:
> * I think this should ignore `<code>` elements.
- +1 re: ignoring `<code>` divs, that's a good catch

> * Should it be considered an error when the contents of an element are a template variable? Example: `ERRO openlibrary/macros/SearchResults.html:23:36  <a href="$book.key/$title.replace(' ', '-')">$book.title_prefix $title</a></span>` or `ERRO openlibrary/templates/covers/book_cover_single_edition.html:18:16  <div class="BookTitle">$:macros.TruncateString(title, 70)`
- I agree that `$`-escaped template variables should be ignored, ideally a `$` that is not followed by a `(` (the syntax intended to trigger `WARN`) should be fine

> * I'm not sure what the meaning is of `\$$`, for example: `ERRO openlibrary/templates/admin/sponsorship.html:69:6  <td>\$$('{:,.2f}'.format(summary['avg_scan_cost'] / 100.))</td>` - is this something else that should avoid flagging?
- I believe that's an escaped `$` associated with a dollar amount (scan cost), a probably very unusual edge case, but we should avoid flagging it

> (I am assuming that this is an indication that the space should be deleted and not that the script should account for this?)
- +1, I think you can go ahead and delete the space! I imagine this script is going to catch a number of random little errors like that which we can just fix

> * I don't have a good answer for tags that are valid inside i18n functions, like `<strong>` or `<em>` or maybe `<span>`. I thought about ignoring any result that matches those three tags, but then that would miss things like `<div><span>untranslated text</span></div>`. I've also considered stripping those tags from the text before processing it, but that would cause incorrect position indexes in the error messages.
> 
- @cdrini has a good solution for this which he can describe in more detail! I believe that the idea is that once you've found a correctly translated string, you check whether any untranslated strings are inside it and ignore those? Or something along those lines, he can go through it fully.

And re: how to just run on changed files, I _believe_ `pre-commit` will automatically just pass changed files to the function, i.e. `i18n_checker.py foo1.html foo2.html` if running following an HTML commit. Again CC @cdrini for how exactly the process works, `pre-commit` itself should be doing most of the heavy lifting for you!

**Comment by pidgezero-one** [comment]
Thank you @rebecca-shoptaw ! I've added some changes that will handle the first four points and will need to think a bit more about how to get the fifth to work.

When I run `pre-commit run` locally, it only runs against staged files, but in the CI pipeline it looks like it's running for all files. I'm not too familiar with the CI pipeline yet so I'm not sure how to fix that!

**Comment by rebecca-shoptaw** [comment]
@pidgezero-one Great, thank you! Definitely worth waiting for Drini's input on #5, he has a solid method he thought through for it some time ago and can easily pass along. Very glad it's running against staged files locally, we can investigate re: the CI.

**Comment by cdrini** [comment]
That all looks good! For 5, I'd say if the line/char with the `>` prefix error has a `$:` somewhere on the line before, then we can skip it. Since anything that contains html will need a `$:` before it (e.g. `$:_('Stay <strong>strong</strong>')`. That should let us avoid these with very few false positives!

**Comment by pidgezero-one** [comment]
Thank you for the guidance @rebecca-shoptaw and @cdrini! 😊 I believe the next challenge is that it flags elements which contain punctuation and no words - do we want to go as far as to treat those as errors (or even warnable if encapsulated in `$()`)?

**Comment by cdrini** [comment]
Wow the results for this are looking so great!! This is going to be so useful :blush: 

Hmmm can you give some examples? I saw a few with `&times;` and `&nbsp;` and `&larr;`. Maybe we can erase these if they're the thing that immediately follow the failing `>`? Something like replace `(&times;|&nbsp;|&larr;)+\s*` after the failing `>` with `''`, and see if it still fails? Would that be too difficult?

**Comment by cdrini** [comment]
Oh also can you try running the script with `time`? I'm curious to see how long it takes!

**Comment by pidgezero-one** [comment]
@cdrini So I think this might actually be a trivial problem to solve with the `\p{L}` operator, but that would require importing the pip `regex` library (the standard `re` lib does not support it). Do you think it'd be worth introducing a new dependency?

I think that access to that operator could be useful to have in general (I've used it in other projects to detect word boundaries for non-romance languages, for example) but wanted to get your (and @rebecca-shoptaw )'s opinions before doing anything drastic.

And when you say running the script with `time`, is that a pre-commit argument or do you mean literally recording timestamps within the script?

**Comment by cdrini** [comment]
Hmm, how would `\p{L}` help in this case?

**Comment by cdrini** [comment]
Oh I just mean run it once with `time` to get an idea of how long it takes; e.g. `time python scripts/detect_missing_i18n.py`

**Comment by pidgezero-one** [comment]
> Hmm, how would `\p{L}` help in this case?

I was thinking that if the `>` is followed only by `[^\p{L}]+` (or HTML entities like &nbsp; I've got a more complicated regex for this in my editor) until the next `<` or newline, it means there are probably no letters contained within the element and shouldn't flag an error.

**Comment by pidgezero-one** [comment]
> Oh I just mean run it once with `time` to get an idea of how long it takes; e.g. `time python scripts/detect_missing_i18n.py`

Thanks! According to pre-commit, here's the results:
```
0.27 real         0.19 user         0.04 sys
```

**Comment by cdrini** [comment]
Nice! That's super speedy 😊 

Ah I see! I think since we'll already need an include-list for the html entities, having just a short include list of punctuation characters should be ok :+1: something like `[(){}[]\/\\:;-]` should cover most of our use cases!

**Comment by cdrini** [comment]
(Although I actually did just recently use `\p` in one of our JS files, so I have a feeling we might indeed need that package in the future! I didn't know that didn't work out-of-the-box in python, TIL!)

**Comment by pidgezero-one** [comment]
Currently I'm detecting HTML entities like `&[a-z]+;` for future-proofing - should I be hardcoding specific entities instead?

**Comment by cdrini** [comment]
Nice! That looks good to me!

**Comment by pidgezero-one** [comment]
My latest commit gets rid of most (but not 100%) false positives, but the list of allowed positive punctuation I found in the results is pretty long, so now I'm getting "line too long" linting errors!

I've left a TODO comment to replace the punctuation arrays with `[^\p{L}]` if we decide to use that library in the future.

I'll take an inventory of the remaining false positives later tonight hopefully!

**Comment by pidgezero-one** [comment]
These are some edge case false positives that I'll list here for posterity and will have to give some more thought to. At least this is going well enough that I can make a comprehensive false positive list!
```
ERRO openlibrary/templates/check_ins/reading_goal_progress.html:22:11  <span class="reading-goal-progress__percentage">($goal.progress%)</span> # A variable encapsulated in punctuation
ERRO openlibrary/templates/showia.html:19:25  <span class="object">ia:$ia</span></td> # looks like this is meant to be a variable or some sort
ERRO openlibrary/templates/showia.html:31:46  <a href="https://archive.org/download/$ia/$(ia)_meta.mrc">https://www.archive.org/download/$ia/$(ia)_meta.mrc</a></span></td> # I could ignore anything that looks like its contents are just a URL
WARN openlibrary/templates/type/list/edit.html:13:5  <h1>$(_("Create a list") if new else _("Edit List"))</h1> # technically it's correct that this is a warning, but why does it use $(_(...?
ERRO openlibrary/templates/type/list/edit.html:34:13  <div class="mia__reorder mia__index">≡ #$(i + 1)</div> # my regex can detect elements containing only punctuation, or elements followed by a variable, but not elements containing only punctuation and variables
ERRO openlibrary/macros/StarRatings.html:48:5  <!-- <span class="star-rating__not-rated">Rate</span> --> # can't detect commented out code!
ERRO openlibrary/macros/StarRatings.html:48:10  <span class="star-rating__not-rated">Rate</span> -->
ERRO openlibrary/templates/account/view.html:50:171  <img class="icon-link__image" src="/static/images/icons/share.svg" width="20"> %s</a>' % share_str # I ignore elements followed by a % variable, but can't detect spaces before it
ERRO openlibrary/templates/merge_request_table/table_row.html:43:53  <span class="commenter">@$c['username']</span> $c['message']</div> # another case of variables mixed in with punctuation
ERRO openlibrary/templates/merge/authors.html:73:21  <span class="ol-num"> ${k} </span>
ERRO openlibrary/macros/FulltextSnippet.html:15:41  <", "&laquo;").replace(">", "&raquo;").replace("{{{", "<mark class='highlight'><strong>").replace("}}}", "</strong></mark>"))&hellip; # this is just straight up code
ERRO openlibrary/templates/type/edition/view.html:454:61  <em>"$work.first_sentence"</em></p> # variable enclosed in punctuation
ERRO openlibrary/templates/books/edit/edition.html:437:29  <!-- <option value="__add__">Add a new ID type</option> -->
ERRO openlibrary/templates/books/edit/edition.html:437:34  <option value="__add__">Add a new ID type</option> -->
ERRO openlibrary/templates/site/banner.html:16:9  <button class="page-banner--dismissable-close">X</button> # clearly meant to be a control
ERRO openlibrary/templates/status.html:48:8  <li> $j : $flags[i][j] </li>
ERRO openlibrary/templates/status.html:73:19  <a href="$pr_status.link">#$pr_status.pull_id: $pr_status.name</a>
ERRO openlibrary/templates/account/readinglog_stats.html:22:13  <span title="First published in $first_publish_year">($first_publish_year)</span>
ERRO openlibrary/templates/covers/add.html:70:21  <input type="url" name="url" id="imageUrl" value="$data.get('url', '')" placeholder="https://..." />
ERRO openlibrary/templates/type/local_id/view.html:35:8  <span class="tag">urn:$page.urn_prefix:{local_id}</span>
ERRO openlibrary/templates/type/author/view.html:85:50  <small>($key)</small></li>
<td class="earnings--$('loss' if scan_cost_liability > 0 else 'profit')">\$$('{:,.2f}'.format(scan_cost_liability / 100.))</td> # I think this one is getting confused by the `>` in the equation
ERRO openlibrary/templates/jsdef/LazyWorkPreview.html:17:17  <span class="first_publish_year">($work['first_publish_year'])</span>
```

**Comment by pidgezero-one** [comment]
Added some improvements! The regexes are getting pretty big and unruly so I've broken them down a bit more.

For elements, it checks that the opening `<` is always followed by `[a-z]` (so as to not confuse it with a `<` character that isn't part of a tag), isn't a `code` or `link` or comment tag, and isn't a closing `/>` or `-->` comment tag. It will ignore any result where `>` is followed by `$:?_?\(`, `$`, `[\r\n<]`, `\$$`, or any combination of: punctuation/spaces, URLs, HTML entities, variables (where variables are `$` followed by any amount of non-spaces, or `%` followed by letters or underscores, or anything encased in `{{` and `}}`).

This doesn't match opening tags followed explicitly by `$:?(`, so that's a separate check to issue a warning instead of an error.

For attributes, it looks for an element (determined the same way as above) but checks if it has a title/alt/placeholder attribute, single or double quoted, that will ignore any result where the contents are empty, or the contents begin with `$:?_?\(` or `\$$`, or if the contents are entirely a combination of: punctuation/spaces, URLs, and variables (where variables are determined the same way as in elements).

This doesn't cover attributes whose values start explicitly with `$:?(`, so that's a separate check to issue a warning instead of an error.

I've also added some output formatting for readability.

The list of false positive examples is down to this, which I've noted with questions:
```
ERRO openlibrary/templates/showia.html:19:25       <span class="object">ia:$ia</span></td> # Is this ia prefix some kind of variable syntax?
ERRO openlibrary/templates/type/local_id/view.html:35:8       <span class="tag">urn:$page.urn_prefix:{local_id}</span> # same as previous
WARN openlibrary/templates/type/list/edit.html:13:5       <h1>$(_("Create a list") if new else _("Edit List"))</h1> # There's quite a few of these, and it is a warning because it starts with $(. Do we need a special case for $(_(? Or perhaps should the warning only trigger on $:?\(['"]
ERRO openlibrary/templates/book_providers/librivox_read_button.html:20:40       <a href="https://librivox.org/">LibriVox</a>. LibriVox is a trusted book provider of public domain audiobooks, narrated by volunteers, distributed for free on the internet.') # This will falsely flag product names like this. Is it acceptable to just put things like this inside $() and leave it as a warning?</li>
```

Other shortcomings:
- The list of punctuation exceptions is reactive and not proactive. It includes special characters like `✓×†♥≡★` simply because those already existed in the code base triggered errors before I added them. In the absence of `[^\p{L}]` support, any future HTML files with different special characters will need to have those characters added to the script in order to pass this pre-commit check.
- It looks like the script is running per-directory and not just once. But the total execution time is still roughly 3 seconds on average.
- Locally, pre-commit will only run against staged files unless --all-files is specified. However, in the CI, it appears to always run against all files. I'm not sure how to fix that.

**Comment by pidgezero-one** [comment]
Added some improvements! The regexes are getting pretty big and unruly so I've broken them down a bit more.

For elements, it checks that the opening `<` is always followed by `[a-z]` (so as to not confuse it with a `<` character that isn't part of a tag), isn't a `code` or `link` or comment tag, and isn't a closing `/>` or `-->` comment tag. It will ignore any result where `>` is followed by `$:?_?\(`, `$`, `[\r\n<]`, `\$$`, or if the tag's entire contents is any combination of: punctuation/spaces, URLs, HTML entities, variables (where variables are `$` followed by any amount of non-spaces, or are formatted like `$xxx(...)` or `$xxx[...]` or ${...},  or `%` followed by letters or underscores, or anything encased in `{{` and `}}`).

This doesn't match opening tags followed explicitly by `$:?(`, so that's a separate check to issue a warning instead of an error.

For attributes, it looks for an element (determined the same way as above) but checks if it has a title/alt/placeholder attribute, single or double quoted, that will ignore any result where the contents are empty, or the contents begin with `$:?_?\(` or `\$$`, or if the contents are entirely a combination of: punctuation/spaces, URLs, and variables (where variables are determined the same way as in elements).

This doesn't cover attributes whose values start explicitly with `$:?(`, so that's a separate check to issue a warning instead of an error.

I've also added some output formatting for readability.

The list of false positive examples is down to this, which I've noted with questions:
```
ERRO openlibrary/templates/showia.html:19:25       <span class="object">ia:$ia</span></td> # Is this ia prefix some kind of variable syntax?
ERRO openlibrary/templates/type/local_id/view.html:35:8       <span class="tag">urn:$page.urn_prefix:{local_id}</span> # same as previous
WARN openlibrary/templates/type/list/edit.html:13:5       <h1>$(_("Create a list") if new else _("Edit List"))</h1> # There's quite a few of these, and it is a warning because it starts with $(. Do we need a special case for $(_(? Or perhaps should the warning only trigger on $:?\(['"]
ERRO openlibrary/templates/book_providers/librivox_read_button.html:20:40       <a href="https://librivox.org/">LibriVox</a>. LibriVox is a trusted book provider of public domain audiobooks, narrated by volunteers, distributed for free on the internet.')</li> # Product names are falsely flagged like this. Is it acceptable to just expect the developer to put such names inside $() and ignore the warning?
ERRO openlibrary/templates/admin/people/view.html:255:1       <pre>$:json_encode(hasattr(person, "dict") and person.dict(), indent="    ", sort_keys=True)</pre> # There are several cases like this where the regex that detects when the element's contents are just $somefunction(...) can't detect nested parentheses. Open to suggestions on this one because I'm kind of stumped on it.
ERRO openlibrary/templates/books/edit.html:27:21       <p class="alert thanks">' + _("Thank you for adding that book! Any more information you could provide would be wonderful!") + "</p>" # Having the _() function mid-element here is an issue, because that function can contain pretty much anything and I don't think I can determine when it ends (and if it's followed by any untranslated text) with regex alone.
ERRO openlibrary/templates/books/edit.html:94:194        <em>OL23919A</em></a>).')</span> # This is clearly some kind of code and not a word.
ERRO openlibrary/macros/WorkInfo.html:16:21        <span class="object">$:'; '.join(thingrepr(i) for i in page[name])</span></td></tr> # Edge case where there's a valid use of a quoted space, but the var detector that starts with $ can't parse it that way. In this situation, would it be reasonable to assign the value of $:'; '.join(thingrepr(i) for i in page[name]) to a variable in the template and rendering that variable inside the element instead?
ERRO openlibrary/templates/books/edit/edition.html:570:57        <em>xii, 346p.</em></div> # This comes from <div class="tip">$_("For example:") <em>xii, 346p.</em></div>. Is this a valid error?
```

Other shortcomings:
- The list of punctuation exceptions is reactive and not proactive. It includes special characters like `✓×†♥≡★` simply because those already existed in the code base and triggered errors before I added them. In the absence of `[^\p{L}]` support, any future HTML files with different special characters will need to have those characters added to the script in order to pass this pre-commit check.
- It looks like the script is running per-directory and not just once. But the total execution time is still roughly 3 seconds on average.
- Locally, pre-commit will only run against staged files unless --all-files is specified. However, in the CI, it appears to always run against all files. I'm not sure how to fix that.

File: .pre-commit-config.yaml
Comment (cdrini) [inline .pre-commit-config.yaml:115]: Ah seems like we need to drop `time` here to make the github CI happy! https://results.pre-commit.ci/run/github/69609/1718565101.yL9bSiQcSAqNRXAwORPvPw Committing this fix to your branch real quick so I can see the output via GH :+1: 

```suggestion
      entry: python ./scripts/detect_missing_i18n.py
```

**Review by cdrini — CHANGES_REQUESTED** [review]
This is looking phenomenal, nice work @pidgezero-one ! A few code fixes :+1:

Also as a next step:

1) Add an EXCLUDE_LIST at the top of this file of all the files that are currently failing
2) Skip over those files when processing, but just output `SKIP {filename} (excluded)`
3) Add a new option to the script, `skip_excluded = True`. This will automatically create the parameter `--no-skip-excluded` which we can call from the command line
4) (me): Create an issue for other folks in the community to start chipping away at the failures, and removing them from the exclude list!

File: scripts/detect_missing_i18n.py
Comment (cdrini) [inline scripts/detect_missing_i18n.py:208]: Oh, a few tweaks on this one!

- We generally use `\t` between columns
- Cool TIL there's an underline terminal code 😁 Can we put that in a helper function to make this a little easier to read?
- We can use ljust to avoid a bit of the math there :+1: 

print('\t'.join([errtype, terminal_underline(filestring).ljust(spacing_base + 12, search_chunk])

**Comment by cdrini** [comment]
To answer your questions! Note I edited your comment to add some labels so I could more easily reference 😁 

(A, `ia:$ia`) Hmm... I think we might need to add an easier way to skip for a given line. Let's add:
    - If the line has `detect-missing-i18n-skip-line` on the line after the failure or if the previous line contains `^ +$?#.*detect-missing-i18n-skip-line`, then we'll just skip the failure and not print out anything. This will give us a way to skip the i18n linting essentially from the html, by specifying e.g. `$# detect-missing-i18n-skip-line` on the line, or on the preceding line.
(B, `$(_('foo') if blah else _('bar')`) If the line contains `_(` after the failing opening bracket, then don't consider it a failure. It already contains i18n of some sort!
(C, LibriVox) Hmmm that looks like an error; that line contains a `$:` before the failing element, so it should not be considered as failing
(D) This seems ok ; this will be labelled by the technique in A
(E) We handle this with (A) or, even better, rejigger the html a bit to be a little cleaner :P
(F) I think that's a valid failure; we _should_ i18n the example in the `<em>`
(G) This looks like it should not fail; the line contains a `$:` before the failing regex match. Same as (C)
(H) That file looks like dead code and needs to be deleted; I'm not even sure if that sytax will compile!


Ok hope that helps! Also, can you expand on this? 

> It looks like the script is running per-directory and not just once. But the total execution time is still roughly 3 seconds on average.

It's running multiple times when triggered via pre-commit?

**Comment by pidgezero-one** [comment]
@cdrini 

> It's running multiple times when triggered via pre-commit?

Seems that way - if I comment out all the lines that print individual errors, it prints out eight summaries, but there's only one loop through `files`:
```
45 files scanned. 72 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
44 files scanned. 65 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
45 files scanned. 39 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
45 files scanned. 62 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
44 files scanned. 40 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
45 files scanned. 32 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
44 files scanned. 63 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
39 files scanned. 85 errors found.
Learn how to fix these errors by reading our i18n documentation: https://github.com/internetarchive/openlibrary/wiki/Internationalization#internationalization-i18n-developers-guide
```

It also looks like the CI results still aren't matching the behaviour of `pre-commit run --all-files` locally. [In the CI, this script is always processing all files (even though it's correctly flagging them as skipped)](https://results.pre-commit.ci/run/github/69609/1719107073.EO5wT4jiQuKh9VlcndQJkA), while locally, `pre-commit run` doesn't do this unless I specify `--all-files`. It also looks like it's exiting with code 1 in the CI, even though `pre-commit run --all-files` exits with code 0 locally. Not sure why this is the case or how to fix that.

Other than that, I've addressed the above comments, and there are no more false positives besides the ones that can be flagged with `detect-missing-i18n-skip-line`! I've added this (both inline and preceding) to a few files as well to verify that it works as expected. The --no-skip-excluded flag seems to work as expected too.

File: .pre-commit-config.yaml
Comment (pidgezero-one) [inline .pre-commit-config.yaml:118]: I've added this so that `SKIP` lines will still show even when the exit code is 0. Pre-commit will hide the script output if the exit code is 0.

File: scripts/detect_missing_i18n.py
Comment (cdrini) [inline scripts/detect_missing_i18n.py:267]:
```suggestion
            else:
```

File: scripts/detect_missing_i18n.py
Comment (cdrini) [inline scripts/detect_missing_i18n.py:259]: Ah this is where the bug is! The errors are being counted even when we _continue_ down below. We should only +1 at the end after we run all the continues.

**Review by cdrini — CHANGES_REQUESTED** [review]
Ok looking good! Ah this _is_ failing locally, the text is just hidden since there's a lot of output. The bug is related to when the errcount is incremented! Should be super close know 😊 

Run `pre-commit run detect-missing-i18n --all-files` then `echo $?` to see the status code.

I'm going to be offline for the next spell but this is the last bug then we should be good to go! Would you mind taking this one over the finish line @rebecca-shoptaw ?

**Comment by pidgezero-one** [comment]
> Run `pre-commit run detect-missing-i18n --all-files` then `echo $?` to see the status code.

I think there's still a discrepancy - I'm also referring to the case where every file is skipped. Up until my most recent changes, `pre-commit run detect-missing-i18n --all-files` skips every file and shows an error code of 0 locally using `echo $?`, but skipping every file in the CI exits with code 1.

**Comment by pidgezero-one** [comment]
Figured it out - the CI is running the script against files I hadn't yet pulled locally and detecting errors in them. I've added all outstanding files to the exclude list so we can start tackling them. All checks pass now!

**Review by cdrini — APPROVED** [review]
Niiiice good thinking! This code lgtm! 😊 @rebecca-shoptaw if you could give it a quick sanity check QA, and then merge, that would be awesome! (I'm afk for a spell 😁)

**Review by rebecca-shoptaw — APPROVED** [review]
@pidgezero-one Great work on this!!
**QA:**
**Tried running `pre-commit run --all-files` with no `i18n` formatting changes:**
✅ Detect missing i18n check passes
❓ Output for all skipped files appears, but if this only happens for `--all-files` which seems to be the case, that's fine with me and could be useful
Screenshot 2024-06-25 at 12 47 08 PM
❓It would be awesome if the skipped files comment was even clearer re: the process to fix the problem(s), i.e. "remove the file from exclude list then run the script again to see what the issue is"

**Tried running `pre-commit` with a test commit after making an `i18n` formatting error in a non-skipped file:**
✅ Successfully got an error message when un-`i18n`-syntaxing both basic text and complex edge cases
✅ Did not see skipped files output

**Tried running `pre-commit` with a test commit after adding new non-`i18n`-syntaxed text:**
✅ Successfully got an error message
✅ Error message disappeared and commit succeeded when I fixed the syntax

Looks great to me! I'm happy to approve a merge, and I'm thinking that once it's merged I'll assign myself a deep dive into those skipped files to try to fix as many as I can, and can bundle a minor instructions clarification into that. 🙂

**Comment by pidgezero-one** [comment]
Thank you Rebecca! 😊

> ❓ Output for all skipped files appears, but if this only happens for `--all-files` which seems to be the case, that's fine with me and could be useful 

Yep, this is the case! Regardless of if a file is skipped or errored, it'll only show if that file was passed to the script, so without --all-files it would only apply to staged files.

>❓It would be awesome if the skipped files comment was even clearer re: the process to fix the problem(s), i.e. "remove the file from exclude list then run the script again to see what the issue is"

By "skipped files comment" what does this refer to? A comment in the code, or the script's output, or the instructions I put in the wiki?

I'll be happy to address that ASAP so it can be merged - unfortunately the longer this PR remains open, the more updates I'll have to make to it as other PRs are merged with HTML changes that will set off the CI 😵‍💫

**Comment by rebecca-shoptaw** [comment]
@pidgezero-one No need to make any more changes! I just meant the comment in the code itself, and I'll tweak it if necessary when I do my new/related PR to fix up those excluded files 🙂 

I've approved the changes, we just need a staff member with merge powers to go ahead and merge!

**Comment by pidgezero-one** [comment]
@rebecca-shoptaw Oh okay cool! Thank you so much! 😄

**Comment by cdrini** [comment]
Woohoo!! Thank you so much folks!!

**Pull request merged**
